File: bhlim/src/Chapter4/DirectedGragh.java
package Chapter4;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class DirectedGragh {
	private Map<Vertex, Set<Vertex>> mMap;

	public DirectedGragh() {
		mMap = new HashMap<Vertex, Set<Vertex>>();
	}

	private boolean addVertex(Vertex vertex) {
		if (!mMap.containsKey(vertex)) {
			mMap.put(vertex, new HashSet<>());
			return true;
		}
		return false;
	}

	public boolean addEdge(Vertex from, Vertex to) {
		if (mMap.containsKey(from)) {
			Set<Vertex> dests = mMap.get(from);
			if (!dests.contains(to)) {
				dests.add(to);
				return true;
			}
		} else {
			addVertex(from);
			addEdge(from, to);
		}

		return false;
	}

	public boolean isConnected(Vertex from, Vertex to) {
		if (mMap.containsKey(from)) {
			Set<Vertex> dests = mMap.get(from);
			return dests.contains(to);
		}

		return false;
	}

	public void printGragh() {
		System.out.println(mMap);
	}
}

File: bhlim/kotlin/lib/src/main/java/com/example/lib/chapter4/DirectedGragh.kt
package com.example.lib.chapter4

import java.util.HashSet
import kotlin.collections.HashMap

data class Vertex(val data: Int) {
    override fun toString(): String {
        return data.toString()
    }

    override fun equals(other: Any?) = data == (other as Vertex).data
    override fun hashCode() = data
}

class DirectedGragh {
    private val mMap: MutableMap<Vertex, MutableSet<Vertex>> = HashMap()

    private fun addVertex(vertex: Vertex): Boolean {
        if (!mMap.containsKey(vertex)) {
            mMap[vertex] = HashSet()
            return true
        }
        return false
    }

    fun addEdge(from: Vertex, to: Vertex): Boolean {
        if (mMap.containsKey(from)) {
            val dests = mMap[from]
            if (dests?.contains(to) == false) {
                dests.add(to)
                return true
            }
        } else {
            addVertex(from)
            addEdge(from, to)
        }

        return false
    }

    fun isConnected(from: Vertex, to: Vertex): Boolean {
        if (mMap.containsKey(from)) {
            val dests = mMap[from]
            return dests?.contains(to) == true
        }

        return false
    }

    fun printGragh() {
        println(mMap)
    }
}